File: backend/app/schemas/jobs.py
"""統一背景任務 (Jobs) 的 API schemas。

聚合來自不同領域的「需要等待的任務」：
- migration:      VM 遷移任務 (vm_migration_jobs)
- script_deploy:  服務模板部署 (script_deploy_logs)
- vm_request:     VM/LXC 開機申請 (vm_requests)
- spec_change:    規格變更申請 (spec_change_requests)

所有來源被正規化到統一的 JobItem 結構，以便前端 Job 中心一致顯示。
"""

from __future__ import annotations

import enum
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class JobKind(str, enum.Enum):
    migration = "migration"
    script_deploy = "script_deploy"
    vm_request = "vm_request"
    spec_change = "spec_change"
    deletion = "deletion"


class JobStatus(str, enum.Enum):
    """正規化狀態。各來源的原始狀態會 map 到此枚舉。"""

    pending = "pending"        # 等待中（尚未開始）
    running = "running"        # 執行中
    completed = "completed"    # 成功完成
    failed = "failed"          # 失敗
    blocked = "blocked"        # 受阻
    cancelled = "cancelled"    # 已取消


# 仍視為「進行中」的狀態（會出現在 active count）
ACTIVE_JOB_STATUSES: set[JobStatus] = {
    JobStatus.pending,
    JobStatus.running,
    JobStatus.blocked,
}


class JobItem(BaseModel):
    """統一的 Job 顯示模型。"""

    id: str = Field(description="複合 ID，格式：<kind>:<source_id>")
    kind: JobKind
    title: str
    status: JobStatus
    progress: int | None = Field(default=None, ge=0, le=100, description="0-100, 若不可估算則為 null")
    message: str | None = None
    user_id: uuid.UUID | None = None
    user_email: str | None = None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    detail_url: str | None = Field(
        default=None,
        description="前端可點擊跳轉的相對路徑（例：/jobs?focus=migration:xxx）",
    )
    meta: dict[str, Any] = Field(default_factory=dict)


class JobsListResponse(BaseModel):
    items: list[JobItem]
    total: int
    active_count: int


class JobDetail(BaseModel):
    """Job 詳細資訊，包含 kind 特定的額外欄位（如腳本輸出、規格 diff 等）。"""

    item: JobItem
    # 各 kind 特定的詳細欄位（前端按 kind 渲染對應區塊）
    output: str | None = Field(
        default=None,
        description="完整文字輸出（script_deploy 的 stdout/stderr；migration 的 last_error）",
    )
    error: str | None = None
    extra: dict[str, Any] = Field(
        default_factory=dict,
        description="kind 特定附加資訊（如 spec_change diff、vm_request 規格、migration 時間戳記）",
    )


__all__ = [
    "ACTIVE_JOB_STATUSES",
    "JobDetail",
    "JobItem",
    "JobKind",
    "JobStatus",
    "JobsListResponse",
]

File: frontend/src/services/jobs.ts
import type { CancelablePromise } from "@/client"
import { OpenAPI } from "@/client"
import { request as __request } from "@/client/core/request"

export type JobKind =
  | "migration"
  | "script_deploy"
  | "vm_request"
  | "spec_change"

export type JobStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "blocked"
  | "cancelled"

export type JobItem = {
  id: string
  kind: JobKind
  title: string
  status: JobStatus
  progress: number | null
  message: string | null
  user_id: string | null
  user_email: string | null
  created_at: string
  updated_at: string
  completed_at: string | null
  detail_url: string | null
  meta: Record<string, unknown>
}

export type JobsListResponse = {
  items: JobItem[]
  total: number
  active_count: number
}

export type JobsListQuery = {
  kinds?: JobKind[]
  statuses?: JobStatus[]
  active_only?: boolean
  limit?: number
  offset?: number
  history_days?: number
}

export type JobDetail = {
  item: JobItem
  output: string | null
  error: string | null
  extra: Record<string, unknown>
}

const csv = (xs: readonly string[] | undefined) =>
  xs && xs.length > 0 ? xs.join(",") : undefined

export const JobsAPI = {
  list(params: JobsListQuery = {}): CancelablePromise<JobsListResponse> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/jobs/",
      query: {
        kinds: csv(params.kinds),
        statuses: csv(params.statuses),
        active_only: params.active_only ? true : undefined,
        limit: params.limit ?? 50,
        offset: params.offset ?? 0,
        history_days: params.history_days ?? 30,
      },
      errors: { 422: "Validation Error" },
    })
  },

  recent(limit = 5): CancelablePromise<JobsListResponse> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/jobs/recent",
      query: { limit },
      errors: { 422: "Validation Error" },
    })
  },

  detail(jobId: string): CancelablePromise<JobDetail> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/jobs/{job_id}",
      path: { job_id: jobId },
      errors: { 403: "Forbidden", 404: "Not found", 422: "Validation Error" },
    })
  },
}

// ─── WebSocket helper ────────────────────────────────────────────────────────

export type JobsSubscriber = (snapshot: JobsListResponse) => void

export function connectJobsWebSocket(
  token: string,
  onSnapshot: JobsSubscriber,
  onError?: (err: Event) => void,
): () => void {
  const proto = window.location.protocol === "https:" ? "wss:" : "ws:"
  const host = window.location.host
  const url = `${proto}//${host}/ws/jobs?token=${encodeURIComponent(token)}`

  let ws: WebSocket | null = null
  let stopped = false
  let reconnectTimer: number | null = null

  const open = () => {
    if (stopped) return
    try {
      ws = new WebSocket(url)
    } catch (err) {
      onError?.(err as Event)
      schedule()
      return
    }
    ws.onmessage = (evt) => {
      try {
        const data = JSON.parse(evt.data) as JobsListResponse
        onSnapshot(data)
      } catch {
        // ignore parse error
      }
    }
    ws.onerror = (err) => onError?.(err)
    ws.onclose = () => {
      ws = null
      schedule()
    }
  }

  const schedule = () => {
    if (stopped) return
    if (reconnectTimer !== null) return
    reconnectTimer = window.setTimeout(() => {
      reconnectTimer = null
      open()
    }, 5000)
  }

  open()

  return () => {
    stopped = true
    if (reconnectTimer !== null) {
      clearTimeout(reconnectTimer)
      reconnectTimer = null
    }
    if (ws) {
      try {
        ws.close()
      } catch {
        // noop
      }
      ws = null
    }
  }
}
